spec/controllers/users_controller_spec.rb

```diff
@@ -0,0 +1,132 @@
+require 'rails_helper'
+
+RSpec.describe "UsersController" do
+include AuthenticationSpecHelper
+  before do
+    @controller = UsersController.new
+  end
+
+  describe 'GET #new' do   
+    context 'Render template new' do
+      it 'renders the `new` template' do
+        get :new
+        expect(response).to render_template('new')
+      end
+    end
+  end
+
+  describe 'POST #create' do  
+    let(:valid_user_params) do
+    {
+      name: 'valid_name',
+      email: 'valid_email@example.com',
+      password: 'valid_password',
+      password_confirmation: 'valid_password',
+      role: :renter ,
+      phone_number: '12345678'
+    }
+    end
+
+    context 'create with valid user' do
+      it 'create a new user' do
+        expect { post(:create, params: { user: valid_user_params }) }.to change(User, :count).by(1)
+        expect(flash[:info]).to eq('Please check your email to activate your account.')
+        expect(response).to have_http_status(303)
+        expect(response).to redirect_to(root_url)
+      end
+    end
+
+    context 'create with invalid user' do
+      it 'case name, email is nil' do
+        post :create, params: { user: attributes_for(:user, email: nil, name: nil) }
+        expect(response).to render_template('new')
+      end
+
+      it 'case email has wrong format' do
+        post :create, params: { user: attributes_for(:user, email: "invalid_email") }
+        expect(response).to have_http_status(422)
+      end
+
+      it "case duplicate emails" do
+        user = create(:user, email: "example@email.com")
+        expect {
+          post :create, params: { user: attributes_for(:user, email: "example@email.com") }
+        }.to_not change(User, :count)
+      end
+
+      it "case password length less than 6" do
+        expect {
+          post :create, params: { user: attributes_for(:user, password:1234) }
+        }.to_not change(User, :count)
+      end
+
+      it "case name length greater than 50" do
+        expect {
+          post :create, params: { user: attributes_for(:user, name:Faker::Lorem.characters(number: 60)) }
+        }.to_not change(User, :count)
+      end
+
+      it "case email length greater than 255" do
+        expect {
+          post :create, params: { user: attributes_for(:user, email:Faker::Lorem.characters(number: 256) + "@example.com") }
+        }.to_not change(User, :count)
+      end
```

Comment:
```suggestion
   let(:response) { post :create, params: request_params }
   let(:request_params) {
     user: invalid_user_params
   }
      shared_examples 'return error' do
        it 'return error status' do
          expect { response }.to have_http_status(expected_http_status)
        end
        it "does not create a resource" do
          expect { response }.to_not change(User, :count)
        end
      end
      context 'with email is nil' do
        it_behaves_like 'return error' do
          let(:expected_http_status) { 422 }
          let(invalid_user_params) { user: attributes_for(:user, email: nil) }
        end
       end
      .....
```
Sử dụng share example để giảm duplicate code 
https://www.netguru.com/blog/eliminate-duplicated-code-rspec-capybara#shared-example